plenum/server/replica.py

```diff
@@ -185,6 +186,75 @@ def clear_below_view(self, view_no):
                 del self._batches[v]
 
 
+class ConsensusDataHelper:
+    def __init__(self, consensus_data: ConsensusSharedData):
+        self.consensus_data = consensus_data
+
+    def preprepare_batch(self, pp: PrePrepare):
+        """
+        After pp had validated, it placed into _preprepared list
+        """
+        if pp in self.consensus_data.preprepared:
+            raise LogicError('New pp cannot be stored in preprepared')
+        if pp in self.consensus_data.prepared:
+            raise LogicError('New pp cannot be stored in prepared')
+        if self.consensus_data.checkpoints and pp.ppSeqNo < self.consensus_data.last_checkpoint.seqNoEnd:
+            raise LogicError('ppSeqNo cannot be lower than last checkpoint')
+        self.consensus_data.preprepared.append(pp)
+
+    def prepare_batch(self, pp: PrePrepare):
+        """
+        After prepared certificate for pp had collected,
+        it removed from _preprepared and placed into _prepared list
+        """
+        if pp not in self.consensus_data.preprepared:
+            raise LogicError('Unprepared pp must be stored in preprepared')
+        self.consensus_data.preprepared.remove(pp)
+        self.consensus_data.prepared.append(pp)
+
+    def clear_batch(self, pp: PrePrepare):
+        """
+        When 3pc batch processed, it removed from _prepared list
+        """
+        if pp in self.consensus_data.preprepared:
+            self.consensus_data.preprepared.remove(pp)
+        if pp in self.consensus_data.prepared:
+            self.consensus_data.prepared.remove(pp)
```

Comment:
Actually according to new logic we don't need this function. We shouldn't clear neither preprepared nor prepared when batch is committed. The only reason to clear them is watermarks update.

Reply:
discussed